fs/operations/check.go

```diff
@@ -358,3 +378,215 @@ func CheckDownload(ctx context.Context, opt *CheckOpt) error {
 	}
 	return CheckFn(ctx, &optCopy)
 }
+
+// CheckSum checks filesystem hashes against a SUM file
+func CheckSum(ctx context.Context, fsrc, fsum fs.Fs, sumFile string, hashType hash.Type, opt *CheckOpt, download bool) error {
+	ci := fs.GetConfig(ctx)
+	progress := false
+	var options CheckOpt
+	if opt != nil {
+		options = *opt
+	} else {
+		// default options for hashsum -c
+		progress = ci.Progress
+		if !progress { // prevent visual clutter
+			options.Combined = os.Stdout
+		}
+	}
+	// CheckSum treats Fsrc and Fdst specially:
+	options.Fsrc = nil  // no file system here, corresponds to the sum list
+	options.Fdst = fsrc // denotes the file system to check
+	opt = &options      // override supplied argument
+
+	if !download && (hashType == hash.None || !opt.Fdst.Hashes().Contains(hashType)) {
+		return errors.Errorf("%s: hash type is not supported by file system: %s", hashType, opt.Fdst)
+	}
+
+	if sumFile == "" {
+		return errors.Errorf("not a sum file: %s", fsum)
+	}
+	sumObj, err := fsum.NewObject(ctx, sumFile)
+	if err != nil {
+		return errors.Wrap(err, "cannot open sum file")
+	}
+	hashes, err := ParseSumFile(ctx, sumObj)
+	if err != nil {
+		return errors.Wrap(err, "failed to parse sum file")
+	}
+
+	c := &checkMarch{
+		tokens:   make(chan struct{}, ci.Checkers),
+		opt:      *opt,
+		progress: progress,
+	}
+	lastErr := ListFn(ctx, opt.Fdst, func(obj fs.Object) {
+		c.checkSum(ctx, obj, download, hashes, hashType)
+	})
+	c.wg.Wait() // wait for background go-routines
+
+	// make census of unhandled sums
+	fi := filter.GetConfig(ctx)
+	for filename, hash := range hashes {
+		if hash == "" { // the sum has been successfully consumed
+			continue
+		}
+		if !fi.IncludeRemote(filename) { // the file was filtered out
+			continue
+		}
+		// filesystem missed the file, sum wasn't consumed
+		err := errors.Errorf("File not in %v", opt.Fdst)
+		fs.Errorf(filename, "%v", err)
+		_ = fs.CountError(err)
+		if lastErr == nil {
+			lastErr = err
+		}
+		atomic.AddInt32(&c.dstFilesMissing, 1)
+		c.reportFilename(filename, opt.MissingOnDst, '+')
+	}
+
+	return c.reportResults(ctx, lastErr)
+}
+
+// checkSum checks single object against golden hashes
+func (c *checkMarch) checkSum(ctx context.Context, obj fs.Object, download bool, hashes HashSums, hashType hash.Type) {
+	remote := obj.Remote()
+	c.ioMu.Lock()
+	sumHash, sumFound := hashes[remote]
+	hashes[remote] = "" // mark sum as consumed
+	c.ioMu.Unlock()
+
+	if !sumFound && c.opt.OneWay {
+		return
+	}
+
+	var err error
+	tr := accounting.Stats(ctx).NewCheckingTransfer(obj)
```

Comment:
@ncw 

After the last _experimental_ commit the accounting for checking transfer starts **here**,
before we call `object.Hash` or spawn a go-routine to hash object contents.
I am not sure what behavior is more correct, both variants feel fine.
**Please advise!**

Sample real-task screenshot with early checker accounting:
```
rclone --exclude sha1.txt sha1sum -C hdd:dumps/sha1.txt hdd:dumps -P --checkers 2 --transfers 2 --download
2021-05-24 18:58:17 NOTICE: = e5-02.bin
2021-05-24 18:58:18 NOTICE: = e5-01.bin
2021-05-24 19:00:31 NOTICE: = e5-03.bin
2021-05-24 19:00:31 NOTICE: = e5-04.bin
Transferred:       19.976Gi / 20.027 GiByte, 100%, 40.205 MiByte/s, ETA 1s
Checks:                 8 / 9, 89%
Transferred:            6 / 8, 75%
Elapsed time:      8m29.5s
Checking:
 *                      e5-07.bin: checking
Transferring:
 *                      e5-05.bin: 99% /2.502Gi, 22.286Mi/s, 0s
 *                      e5-06.bin: 98% /2.504Gi, 22.098Mi/s, 1s
```

Reply:
I still don't like current pacer. In case of many files it _walks_ next --checkers files, calculates ETA, then adds next checker, updates ETA and so on. This way ETA is all way small, decreases to almost zero, then jumps to another small duration, then again. This is useless. I would like it to behave like `rclone sync` - count number (or sum sizes in case of --download) of as many files as possible during _walk_. I don't know how to modify existing code to get there. Please help, @ncw.

Reply:
In an ideal world we'd have local checkers counting transfers like downloads do - there have been loads of requests for that.

This could use a similar mechanism to this

https://github.com/rclone/rclone/blob/df60e6323c07a20de08c73839c67aa5f44e141e4/fs/accounting/accounting.go#L275-L295

If checksums for the local backend always used this mechanism then it would solve the UI problem were users are wondering why their B2/S3 uploads don't start because the local backend is making the checksum in advance...

> I still don't like current pacer. In case of many files it walks up next --checkers files, calculates ETA, then adds next checker, updates ETA and so on. This way ETA is pretty small all the time seeing only partial work, then steadily decreases close to zero, then jumps back to another small duration, then again. This is useless in practice.

To achieve this we'd want to move some of the logic from fs/sync into fs/march or possibly run `rclone check` from the `rclone sync` framework rather than using fs/march.

The logic missing is queuing the `checks` so we have an idea of how many we need to do which is only in `fs/sync`.

Anyway I think any of these suggestions should probably go in a new PR

Reply:
> I think any of these suggestions should probably go in a new PR.

Totally agree. This is too tricky!

Reply:
Please let me return to this problem when get to self-assigned #2749

Reply:
I saved your suggestions at https://github.com/rclone/rclone/issues/2749#issuecomment-850789836

Reply:
Let me return to check accounting later in #2749

Reply:
Let me return to the check accounting problem later in #2749

Reply:
Let me return to the check accounting problem later in #2749 :pray: